Automatically detect whether the terminal background is a light color

eschnett opened the pull request:
This detects (or tries to) whether the terminal background is a light color. The REPL then chooses a different default color scheme that is easier to read, avoiding e.g. yellow.

This only affects the default colors; any user-chosen colors remain the same.

Closes #28690.

KristofferC commented:
What is the user has changed the way the default colours are shown in the terminal to already fix this issue? Then we will still swap to these non-standard colors. 

If the defaults in some terminals make some colors unreadable, maybe it is the defaults that need to be changed?

Keno commented on stdlib/REPL/src/Terminals.jl:
```diff
@@ -169,6 +171,31 @@ else
+    # (We could/should use more mechanisms here)
+    # Blindly guess an OS-dependent default
+    if Sys.isapple()
+        return true
```
This is a bad assumption. iTerm2 defaults to black. However, iTerm2 does have color query escape codes.

eschnett replied:
To my knowledge, iTerm2 sets the `COLORFGBG` environment variable that we check above, thus this default isn't reached.

KristofferC replied:
I don't think we should ever default this to true solely based on OS.

eschnett replied:
I removed the default based on OS; all OS now use the same default

eschnett commented:
@KristofferC If the user sets colors, then Julia won't use its default colors. This change only modifies the defaults that Julia uses if there is no customization.

I don't think it is possible to choose colors that work with every background. Light colors are unreadable on light backgrounds, and dark colors are unreadable on dark backgrounds. Plus, Julia wants to use several different colors to improve the output.

StefanKarpinski commented:
This seems like a good direction to make the default experience better. But does this light/dark background detection really work? It seems kind of implausible to me that it would actually work.

eschnett commented:
Appveyor reports an all-green status, but the Github status here on this page doesn't update.

eschnett committed:
Allow both "light" and "dark" text colors

Introduce a distinction between light and dark text colors. Use different default colors in both cases.

The text color scheme can be set via a command line option (--color=[light | dark]), or Julia can try to detect the terminal's background color. In this case, it roughly follows vim's approach:
1. Look for COLORFGBG environment variable
2. Use an ANSI escape sequence to query the terminal for the background color
3. Make an educated guess from the terminal name

Tested with iTerm2, Terminal, xterm on MacOS, Linux.

eschnett commented:
The patch now uses roughly the same mechanism as vim. Querying the terminal via ANSI escape sequences work for iTerm2 and xterm (and probably their derivatives). There are also command line options to override things (`--color=[auto | on | off | light | dark]`). Ready for review.

eschnett committed:
Invert ANSI colors

eschnett committed:
Merge branch 'master' into eschnett/light-background

eschnett committed:
Experiment with colours

StefanKarpinski commented:
Bump, who should review?

eschnett commented:
I now find that many Julia packages set colours directly (e.g. by using `:yellow`), and are not going through either environment variables or a colour scheme (as the REPL does). It seems that it would be a huge task to change this, as it would involve changing many packages as well (Pkg, profiler, ..., ?), and maybe even setting up usability coding guidelines. (Or maybe Julia should set the terminal background colour as well, to ensure that everybody is using the same background colour and thus has readable colours.)

A work-around would be to conditionally change the definition of `:yellow` to be `:orange` instead, i.e. to change the meaning of colours behind package writers' back. That goes in the same direction as @KristofferC's comment, except that he changes the terminal settings.

In summary, I'm disappointed that my pull request didn't actually solve the problem it was supposed to solve. I don't know whether it's actually useful in practice. If so, then it's only a first step towards making Julia output readable on a light background.

vtjnash commented on stdlib/REPL/src/Terminals.jl:
```diff
@@ -169,7 +171,101 @@ else
+    if term.term_type == "linux" ||         # Linux console
+        term.term_type == "screen.linux" || # Linux console with screen
+        startswith(term.term_type, "cygwin") || # Cygwin shell
+        startswith(term.term_type, "putty")     # Putty progrem
```
I know this list comes from VIM, but it seems likely ancient. I think most terminals I use report as "xterm" now

eschnett replied:
This is a list of terminals which default to using a dark background. The terminals which default to a light background (`xterm` and friends) are not in this list.

eschnett commented on stdlib/REPL/src/Terminals.jl:
```diff
@@ -169,7 +171,101 @@ else
+        end
+    end
+
+    # Use OSC 11 escape sequence
```
Leave out which part – looking at the `COLORFGBG` variable?

vtjnash commented:
I think we've somewhat essentially decided not to go with this PR. I think the `--color=light` / `--color=dark` options still have merit however, so I'll keep the corresponding issue open and mention that there

StefanKarpinski commented:
Probably better as an environment variable since you presumably don't want to have to set it on each invocation.
